cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -2644,7 +2645,13 @@ private ApplicationMeta decodeRow(StructuredRow row) {
     } else {
       changeDetail = new ChangeDetail(changeSummary, null, author, creationTimeMillis, latest);
     }
-    return new ApplicationMeta(id, spec, changeDetail, sourceControl, latest);
+
+    return new ApplicationMeta(id, spec, changeDetail, sourceControl,
+        isLatest || Boolean.TRUE.equals(latest));
+  }
+
+  private ApplicationMeta decodeRow(StructuredRow row) {
```

Comment:
looking at this class a bit more carefully, it looks like this method is called by `AppMetadataStore.getApplication(ApplicationId)`. In that case, when isLatest is missing, this method will return ApplicationMeta with latest set to false, which can be incorrect.

It looks like this flag is never used/needed when reading a row -- it is only ever used in AppMetadataStore.createApplicationVersion() to indicate whether we need to set the flag when inserting a new row. If that's the case, it would be better to refactor the createApplicationVersion() method to take it as a separate argument rather than encapsulating it within the ApplicationMeta class.

If that would be too much refactoring right now, it would be ok to temporarily make it a nullable Boolean in ApplicationMeta with a TODO jira to complete the refactor.

Reply:
I think we can refactor the `createApplicationVersion` method and remove the `transient boolean isLatest` from `ApplicationMeta`. I had added this change in this [PR](https://github.com/cdapio/cdap/pull/15336/files#diff-7606ad53d616b95330a3ca367c096517fc8945694e053d7e2369fcea59ba54c4) to avoid overloading the `createApplicationVersion` method (based on the discussion [here](https://github.com/cdapio/cdap/pull/15336#discussion_r1345170315)).

Though, with this refactor we will end up with one more overloaded `addApplication` method in the `Store` interface.

Reply:
@albertshau pushed the refactor as a new commit in this PR. Should we squash these commits or keep them separate ?